### ISSUE TITLE
Error in autosubmit run  using 3.12.1b

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#403]
In GitLab by @f-macchia on May 28, 2019, 16:06

I'm testing the last beta version of autosubmit (3.12.1b) running exp a1yg. The `autosubmit run` crash after some hours with this error:

```
[ERROR] Can not send file /esarchive/autosubmit/a1yg/tmp/a1yg_20110101_003_281_REDUCE.cmd to /gpfs/scratch/bsc32/bsc32557/a1yg/LOG_a1yg/a1yg_20110101_003_281_REDUCE.cmd                            
[ERROR] power9 submission failed
[CRITICAL] Unhandled exception on Autosubmit: size mismatch in put!  0 != 6269
Traceback (most recent call last):                                                                
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 418, in parse_args              
    return Autosubmit.run_experiment(args.expid, args.notransitive)                               
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 1076, in run_experiment
    if Autosubmit.submit_ready_jobs(as_conf, job_list, platforms_to_test, packages_persistence):
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 1123, in submit_ready_jobs
    package.submit(as_conf, job_list.parameters, inspect)
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/job/job_packages.py", line 92, in submit
    self._send_files()
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/job/job_packages.py", line 122, in _send_files
    self.platform.send_file(self._job_scripts[job.name])
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/platforms/paramiko_platform.py", line 135, in send_file
    self._ftpChannel.put(os.path.join(self.tmp_path, filename), os.path.join(self.get_files_path(), os.path.basename(filename)))
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/paramiko-1.15.0-py2.7.egg/paramiko/sftp_client.py", line 669, in put
    return self.putfo(fl, remotepath, file_size, callback, confirm)
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/paramiko-1.15.0-py2.7.egg/paramiko/sftp_client.py", line 635, in putfo
    raise IOError('size mismatch in put!  %d != %d' % (s.st_size, size))
IOError: size mismatch in put!  0 != 6269
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 31, 2019, 16:54

Hi @f-macchia,

These crashes are happening always in the same job? I have seen this error before but can't remember  now where

Related to #394, since both are I/O errors

I'll try to implement this change commented by @mcastril:

> @dbeltrankyl , maybe instead of taking a lot of things into consideration to detect platform availability, could we implement a way to recover Autosubmit status when an exception is thrown? I mean, once the exception is captured at least we know there's a problem. Maybe in that case Autosubmit can be re-run as if someone types Autosubmit run. Of course this would only be valid for connection or I/O issues and not othe kind of exceptions. 


I think that would solve a large part of the enza frozen experiment errors and this one

Edit: was #394 where was reported the error I just saw it

Edit2: Both were in the reduce.cmd although different members and chunk

Edit3: Another one, it was happening also in 3.12.0b, is weird because the cmd is not corrupted in local so, it should be able to send it, seems like the related issue, a power9 problem. @mcastril proposed solution should work once implemented, likely in Monday-Tuesday I'll update the version.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 3, 2019, 10:58

Seems like a platform specific failure right? Paramiko with Power9

When you develop this new feature (exceptions handling), I am thinking that we can save the error in a way that we can afterwards know which kind of problems Autosubmit is having and in which kind of jobs.

More concretely I am thinking about generating a new file (error file with a timestamp in the filename) with information inside about the error message, the job in which it was produced, platform, etc.

It could work to gather statistics and report Operations about problems we are having with Power9 or other platforms.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 3, 2019, 17:57

Hi @f-macchia @mcastril ,

I made some changes(and updated 3.12.1b) to how paramiko handle the execution of submit and check commands, It seems that sometimes the function that checks if the command has finished in remote, hangs autosubmit so I used another function, make my own timeout and change the mechanism to a pool one with timeouts(30s) and retries (5 times).

If there is still no response from the platform autosubmit now it just will handle the IO exception and the timeout exception and skip the package or platform (depending on if it is checking the status or is submitting) until the next iteration of autosubmit run.

If AS Fails to check job status it will change into unknown status and in the next iteration, it will check and set if the job is completed or failed.

If AS fails in the submitting function there won't be consequences since AS won't submit anything. This means that the job won't have an id and thus will remain in ready status for the next iteration.

About the logs, I changed some of the logs info but, in order to make them in a separated new file I would need to do the log in a different way than the actual ones since the library only allows to set one log file, also I'm trying to solve how to properly store all exceptions into a file.
